### PR TITLE
mam2: add and use troika

### PR DESCRIPTION
**Pull request by tsvisabo**
- Integrate Troika from https://github.com/Troikahash/reference
- Use Troika for transformation

### CODE REVIEW

File: common/troika/troika.c [hunk 1]
```diff
@@ -0,0 +1,498 @@
+/*
+* Copyright (c) 2019 IOTA Stiftung
+* Copyright (c) 2019 Cybercrypt A/S
+ *
```
Comment (thibault-martinez) [inline common/troika/troika.c:5]: Remove space

File: common/troika/troika.c [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+* https://github.com/iotaledger/entangled
+*
+Permission is hereby granted, free of charge, to any person obtaining a copy
```
Comment (thibault-martinez) [inline common/troika/troika.c]: Can you add `* ` on every line like all our copyrights ?

File: common/troika/troika.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
+/*
+* Copyright (c) 2019 IOTA Stiftung
+* Copyright (c) 2019 Cybercrypt A/S
+ *
```
Comment (thibault-martinez) [inline common/troika/troika.h:5]: Same here

File: common/troika/troika.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+* https://github.com/iotaledger/entangled
+*
+Permission is hereby granted, free of charge, to any person obtaining a copy
```
Comment (thibault-martinez) [inline common/troika/troika.h]: Same here

File: mam/v2/sponge/sponge.c [hunk 5]
```diff
@@ -70,21 +71,7 @@ void mam_sponge_init(mam_sponge_t *const sponge) {
 
 // TODO mam2: replace sponge transformation function #773
```
Comment (thibault-martinez) [inline mam/v2/sponge/sponge.c]: Remove  comment

File: mam/v2/troika/troika.c [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * [ITSec Lab]
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include "mam/v2/troika/troika.h"
+#include "common/troika/troika.h"
+#include "mam/v2/trits/trits.h"
+
+/*void troika_permutation(trit_t *state, unsigned long num_rounds)*/
```
Comment (thibault-martinez) [inline mam/v2/troika/troika.c]: Remove ?

File: mam/v2/troika/troika.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ * Refer to the LICENSE file for licensing information
+ */
+
+#ifndef __MAM_V2_TROIKA_H__
+#define __MAM_V2_TROIKA_H__
+
+#include "common/troika/troika.h"
+#include "mam/v2/defs.h"
+
+/*#define MAM2_TROIKA_NUM_ROUNDS 12*/
```
Comment (thibault-martinez) [inline mam/v2/troika/troika.h:16]: Remove ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: common/troika/troika.h (hunk 3 above)
Comment (thibault-martinez) [reply common/troika/troika.h:5]: Hasn't been removed :(

File: common/troika/troika.h (hunk 4 above)
Comment (thibault-martinez) [reply common/troika/troika.h]: Hasn't been added :(

File: mam/v2/troika/troika.h (hunk 7 above)
Comment (thibault-martinez) [reply mam/v2/troika/troika.h:16]: Not removed :(

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — DISMISSED** [review]

File: mam/v2/mss/tests/test_mss.c [hunk 8]
```diff
@@ -12,6 +12,7 @@
 
 #include "mam/v2/mss/mss.h"
 #include "mam/v2/test_utils/test_utils.h"
+#include "utils/handles/rand.h"
```
Comment (thibault-martinez) [inline mam/v2/mss/tests/test_mss.c:15]: Why ?

File: mam/v2/mss/tests/BUILD [hunk 9]
```diff
@@ -5,6 +5,7 @@ cc_test(
     deps = [
         "//mam/v2/mss",
         "//mam/v2/test_utils",
+        "//utils/handles:rand",
```
Comment (thibault-martinez) [inline mam/v2/mss/tests/BUILD:8]: Why ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/v2/mss/tests/test_mss.c (hunk 8 above)
Comment (tsvisabo) [reply mam/v2/mss/tests/test_mss.c:15]: Fixed with usage

File: mam/v2/mss/tests/BUILD (hunk 9 above)
Comment (tsvisabo) [reply mam/v2/mss/tests/BUILD:8]: Fixed with usage

File: mam/v2/mss/tests/test_mss.c [hunk 10]
```diff
@@ -204,6 +204,9 @@ static bool mss_test(mss_t *mss, mam_prng_t *prng, mam_spongos_t *spongos,
     mss_gen(mss, pk);
 
     do {
+      if (curr_height > 1 && ((rand() % 2) == 0)) {
```
Comment (thibault-martinez) [inline mam/v2/mss/tests/test_mss.c]: Use `rand_handle_rand`

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Comment by Citrullin** [comment]
@thibault-martinez @tsvisabo Is it clear that the reference implementation is not supposed to be used in production like environments? There are some problems with the reference implementation. E.g. [this one.](https://github.com/Troikahash/reference/issues/2)
The performance of the reference implementation is also not really nice. One IOTA community dev already improved the implementation a lot. You can find the repo [here](https://github.com/c-mnd/troika). It is ~10 times faster than the reference implementation, and he did fix the squeeze absorb bug.

**Comment by shufps** [comment]
for fixed absorb this [here](https://github.com/c-mnd/troika/pull/1). It hasn't been merged in the optimized troika version yet ... I'm not sure if it will happen since the creator of the optimizations is anonymous and he only created the github repo (and his github-account) only for troika.

**Pull request merged**